rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/ClassFileListener.java

```diff
@@ -0,0 +1,201 @@
+package org.jboss.windup.rules.apps.java;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.bcel.classfile.ClassParser;
+import org.apache.bcel.classfile.ConstantClass;
+import org.apache.bcel.classfile.DescendingVisitor;
+import org.apache.bcel.classfile.EmptyVisitor;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.commons.lang.StringUtils;
+import org.jboss.windup.config.GraphRewrite;
+import org.jboss.windup.graph.model.resource.FileModel;
+import org.jboss.windup.graph.service.GraphService;
+import org.jboss.windup.reporting.service.ClassificationService;
+import org.jboss.windup.rules.apps.java.model.JavaClassFileModel;
+import org.jboss.windup.rules.apps.java.model.JavaClassModel;
+import org.jboss.windup.rules.apps.java.scan.ast.TypeInterestResolver;
+import org.jboss.windup.rules.apps.java.service.JavaClassService;
+import org.jboss.windup.rules.apps.java.service.WindupJavaConfigurationService;
+import org.jboss.windup.rules.files.FileDiscoveredEvent;
+import org.jboss.windup.rules.files.FileDiscoveredListener;
+import org.jboss.windup.rules.files.FileDiscoveredResult;
+import org.jboss.windup.util.ExecutionStatistics;
+import org.jboss.windup.util.Logging;
+import org.objectweb.asm.ClassReader;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+/**
+ * @author <a href="mailto:jesse.sightler@gmail.com">Jesse Sightler</a>
+ */
+public class ClassFileListener implements FileDiscoveredListener
+{
+    private static Logger LOG = Logging.get(ClassFileListener.class);
+
+    @Override
+    public void fileModelCreated(GraphRewrite event, EvaluationContext context, FileModel fileModel)
+    {
+        ExecutionStatistics.get().begin("ClassFileListener.fileModelCreated()");
+        try
+        {
+            if (fileModel.isDirectory())
+                return;
+
+            if (!isClassFile(fileModel.getFileName()))
+                return;
+
+            try (InputStream is = fileModel.asInputStream())
+            {
+                ClassReader classReader = new ClassReader(is);
```

Comment:
this line would be probably better to do after we know for sure that we are going to look inside (it doesn't matter in case classReader does not have any performace-intensive initialization).